python_modules/dagster/dagster/components/core/load_defs.py

```diff
@@ -26,12 +27,42 @@ def build_component_defs(components_root: Path) -> Definitions:
         f"{Path(components_root).parent.name}.{Path(components_root).name}"
     )
 
-    return load_defs(defs_root=defs_root)
+    return load_defs(defs_root=defs_root, project_root=components_root.parent.parent)
+
+
+def get_project_root(defs_root: ModuleType) -> Path:
+    """Find the project root directory containing pyproject.toml or setup.py.
+
+    Args:
+        defs_root: A module object from which to start the search.
+
+    Returns:
+        The absolute path to the project root directory.
+
+    Raises:
+        FileNotFoundError: If no project root with pyproject.toml or setup.py is found.
+    """
+    # Get the module's file path
+
+    module_path = getattr(defs_root, "__file__", None)
+    if not module_path:
+        raise FileNotFoundError(f"Module {defs_root} has no __file__ attribute")
+
+    # Start with the directory containing the module
+    current_dir = Path(module_path).parent
+
+    # Traverse up until we find pyproject.yaml or setup.py
+    while current_dir != current_dir.parent:  # Stop at root
+        if (current_dir / "pyproject.toml").exists() or (current_dir / "setup.py").exists():
+            return current_dir
+        current_dir = current_dir.parent
+
+    raise FileNotFoundError("No project root with pyproject.toml or setup.py found")
 
 
 # Public method so optional Nones are fine
 @suppress_dagster_warnings
-def load_defs(defs_root: ModuleType) -> Definitions:
+def load_defs(defs_root: ModuleType, project_root: Optional[Path] = None) -> Definitions:
```

Comment:
We can change the project scaffolder to use this argument. I think that would be more explicit and less error-prone. Followup?